case 21983: Clear more disk caches

orenh1 opened the pull request:
When the user selects "Clear Disk Cache" from the Developer menu, clear more of the caches.

https://highfidelity.fogbugz.com/f/cases/21983/Clear-Disk-Cache-menu-command-doesn-t-clear-all-caches

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18441/HighFidelity-Beta-PR15278-c08f2cb.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18441/HighFidelity-Beta-Interface-PR15278-c08f2cb.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18441/HighFidelity-Beta-PR15278-c08f2cb.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18441/HighFidelity-Beta-Interface-PR15278-c08f2cb.exe

sethalves commented:
Android build failed like this:

14:12:01   /home/jenkins/hifi/interface/src/Menu.cpp:687:9: error: use of undeclared identifier 'FileTypeProfile'
14:12:01           FileTypeProfile::clearCache();
14:12:01           ^
14:12:01   /home/jenkins/hifi/interface/src/Menu.cpp:688:9: error: use of undeclared identifier 'HFWebEngineProfile'
14:12:01           HFWebEngineProfile::clearCache();
14:12:01           ^
14:12:01   2 errors generated.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18447/HighFidelity-Beta-PR15278-571fbf9.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18447/HighFidelity-Beta-Interface-PR15278-571fbf9.dmg

hifi-gustavo commented:
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18447/HighFidelity-Beta-PR15278-571fbf9.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18447/HighFidelity-Quest-Beta-PR15278-571fbf9.apk)

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18447/HighFidelity-Beta-PR15278-571fbf9.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18447/HighFidelity-Beta-Interface-PR15278-571fbf9.exe

AndrewMeadows commented on libraries/ui/src/ui/types/FileTypeProfile.cpp:
```diff
@@ -37,5 +43,10 @@ void FileTypeProfile::registerWithContext(QQmlContext* context) {
 
+void FileTypeProfile::clearCache() {
+    if (!instance)
+        return;
```
Coding standard says the `if` block needs braces like this:
```
if (!instance) {
    return;
}
```
Unfortunately I can't find a valid link to the coding standard right now (I will complain internally) so in the meantime you'll just have to take my word for it.

AndrewMeadows replied:
Or better yet avoid the negation/early-exit and just do:
```
if (instance) {
    instance->clearHttpCache();
}
```

AndrewMeadows commented on libraries/ui/src/ui/types/FileTypeProfile.h:
```diff
@@ -25,13 +25,17 @@ class FileTypeProfile : public ContextAwareProfile {
         void interceptRequest(QWebEngineUrlRequestInfo& info) override;
     };
+
+    static FileTypeProfile* instance;
```
Not a blocker but my own preference: `instance` needs to be declared in the header because there is no need (yet) to expose it with `protected` status.  Just declare it in the CPP file but give it a less namespace-clashy name like `FileTypeProfile_instance` or something.

AndrewMeadows commented on libraries/ui/src/ui/types/HFWebEngineProfile.cpp:
```diff
@@ -35,4 +39,10 @@ void HFWebEngineProfile::registerWithContext(QQmlContext* context) {
 
+void HFWebEngineProfile::clearCache() {
+    if (!instance)
+        return;
```
The logic block inside all `if` statements must be enclosed with braces, even when they are one line.

AndrewMeadows commented on libraries/ui/src/ui/types/HFWebEngineProfile.h:
```diff
@@ -23,13 +23,17 @@ class HFWebEngineProfile : public ContextAwareProfile {
         void interceptRequest(QWebEngineUrlRequestInfo& info) override;
     };
+
+    static HFWebEngineProfile* instance;
```
Not a blocker but: `instance` does not need to be exposed with `protected` status.  I recommend moving it to the CPP file.

AndrewMeadows commented:
Our internal QA + ProjectManagement process requires an internal "case" associated with this PR.  I will create one and hook it up, however it could use a simple test recipe for verifying the changes this PR introduces.  For example, something like:

**(1)** Before starting interface look in Windows directory `some\path` and notice all the cache files
**(2)** Start interface
**(3)** Clear cache via **this menu option**
**(4)** Look in Windows directory `some\path` and notice the cache files are no longer there
**(5)** Try on an older version of the interface and notice the cache files are not removed.

If you could provide those instructions I'll add them to the case.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18504/HighFidelity-Beta-PR15278-fe10f3f.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18504/HighFidelity-Beta-Interface-PR15278-fe10f3f.dmg

hifi-gustavo commented:
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18504/HighFidelity-Beta-PR15278-fe10f3f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18504/HighFidelity-Quest-Beta-PR15278-fe10f3f.apk)

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18504/HighFidelity-Beta-PR15278-fe10f3f.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18504/HighFidelity-Beta-Interface-PR15278-fe10f3f.exe

orenh1 commented:
Test plan:

1. Start Interface and visit a couple of domains. Also, switch between a few avatars. The purpose of this step is to fill the caches.
2. Restart Interface
3. Identify your Local Data directory. On Windows, it should have a path similar to: C:\Users\My_Name\AppData\Local\High Fidelity\Interface
4. Using Windows Explorer, find and write down the size of each of the following directories under your Local Data directory: cache, ktx_cache
5. Enable the Developer menu, and select the menu item: Developer > Network > Clear Disk Cache
6. Check the new size of each of these directories. It should be smaller than before. (But not necessarily 0, because files that are currently in use aren't deleted.)
7. Repeat this procedure with an older version of Interface, and note that the size of these directories isn't reduced when you select Clear Disk Cache.

samcake commented:
IN theory the KTX cache can NOT be cleared while we are using the application because potentially some files are in use.
How does this pr addresses this ?

samcake commented on libraries/material-networking/src/material-networking/TextureCache.cpp:
```diff
@@ -83,6 +83,11 @@ TextureCache::TextureCache() {
+void TextureCache::wipe() {
+    clearUnusedResources();
+    _ktxCache->wipe();
+}
```
The ktxCache is not designed to be cleared at runtime because some textures are potentially in use already and assume that their data is available from files in that cache.
The way it s cleared right now is done from the Developer/Render/clear ktx cache menu option, it requires to shut down the app and then is executed on the next run of the app.

samcake commented:
I would suggest to modify the current pr without clearing the ktx cache for now

orenh1 commented:
How about we clear the KTX cache on next start, like the existing "Clear KTX Cache" option?

samcake commented:
At this point, i would recommend to postpone to 84, sorry too much to deal with right now.
Yes the idea of using the similar clear ktx cache as already existing sounds correct. But we don;t need the new  "wipe cache" calls introduced in this pr, instead go through the exact same call path.
Thanks

orenh1 commented:
I heard you guys were asking about the status of this PR in the developer meeting on July 10. Sorry I haven't updated it yet: I work on multiple fronts, and I haven't been working on the viewer front lately. But now that I know people are actually waiting for this I'll add an update soon.

BTW, in that meeting I was described as the main guy in Kitely. I'm the main developer, but I have a partner, Ilan Tochner, who's our CEO. He just doesn't write pull requests so you haven't seen him here...

orenh1 commented:
I have updated this PR:

1. I renamed the menu item to "Clear Disk Caches (requires restart)"
2. The KTX cache is only cleared after a restart
3. Due to Qt changes, QtWebEngine was no longer storing its cache on disk (only in memory). I have fixed this.
4. My previous practice of finding just one instance of each type of cache and cleaning it doesn't work. I'm not sure if this is also due to Qt changes, or if it has only worked in the past by chance. In any case, now I keep track of all of the instances of each cache type, and clean all of them.

orenh1 commented:
For reference, here are some of the cache directories and which class is responsible for them:

cache/QtWebEngine - QtWebEngine cached files (HTML, images, etc.). Cleared using clearHttpCache().
QtWebEngine - QtWebEngine persistent storage (e.g., cookies). There's no API to clear this.
data8 - Cleared using AssetClient.clearCache()
ktx_cache - Cleared using KTXCache.wipe()

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20599/HighFidelity-Beta-PR15278-d7ef7a0.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/20599/HighFidelity-Beta-Interface-PR15278-d7ef7a0.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20599/HighFidelity-Beta-PR15278-d7ef7a0.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/20599/HighFidelity-Beta-Interface-PR15278-d7ef7a0.dmg

hifi-gustavo commented:
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20599/HighFidelity-Beta-PR15278-d7ef7a0.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/20599/HighFidelity-Quest-Beta-PR15278-d7ef7a0.apk)
